clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/NettyP2PBlobTransferManager.java

```diff
@@ -41,132 +57,143 @@ public NettyP2PBlobTransferManager(
     this.blobTransferService = blobTransferService;
     this.nettyClient = nettyClient;
     this.peerFinder = peerFinder;
+    this.baseDir = nettyClient.getBaseDir();
   }
 
   @Override
   public void start() throws Exception {
     blobTransferService.start();
   }
 
+  @Override
+  public CompletionStage<InputStream> get(String storeName, int version, int partition)
+      throws VenicePeersNotFoundException {
+    CompletableFuture<InputStream> resultFuture = new CompletableFuture<>();
+    // 1. Discover peers for the requested blob
+    BlobPeersDiscoveryResponse response = peerFinder.discoverBlobPeers(storeName, version, partition);
+    if (response == null || response.isError() || response.getDiscoveryResult() == null
+        || response.getDiscoveryResult().isEmpty()) {
+      // error case 1: no peers are found for the requested blob
+      String errorMsg = String.format(NO_PEERS_FOUND_ERROR_MSG_FORMAT, storeName, version, partition);
+      resultFuture.completeExceptionally(new VenicePeersNotFoundException(errorMsg));
+      return resultFuture;
+    }
+
+    List<String> discoverPeers = response.getDiscoveryResult();
+    LOGGER
+        .info("Discovered peers {} for store {} version {} partition {}", discoverPeers, storeName, version, partition);
+
+    // 2: Process peers sequentially to fetch the blob
+    processPeersSequentially(discoverPeers, storeName, version, partition, resultFuture);
```

Comment:
Why not just return the future from this method so we can eliminate the additional argument?